src/storage/redis_storage.rs

```diff
@@ -1,34 +1,247 @@
+        pool: Pool<RedisConnectionManager>,
+        ttl_in_seconds: usize,
+        max_command_messages: isize,
+    ) -> Result<Self> {
+        Ok(Self {
+            prefix: prefix,
+            pool: pool,
+            ttl_in_seconds: ttl_in_seconds,
+            max_command_messages: max_command_messages,
+        })
+    }
+
+    // TODO: add some methods to for common redis commands
+    // instead of repeating it multiple times
+
+    pub async fn get_connection(&self) -> Result<PooledConnection<'_, RedisConnectionManager>> {
+        let conn = self
+            .pool
+            .get()
+            .await
+            .context("unable to retrieve a redis connection from the pool")?;
+
+        Ok(conn)
+    }
+}
+
+// TODO: anything that can accept String and &str we can use Into<String>?
+// TODO: what about losing messages while moving them between queues?
 
 #[async_trait]
 impl Storage for RedisStorage {
-    async fn get(&self, _id: String) -> anyhow::Result<Option<Message>> {
-        unimplemented!()
+    async fn get(&self, id: String) -> Result<Option<Message>> {
+        let mut conn = self.get_connection().await?;
+        let key: String = format!("{}.backlog.{}", &self.prefix, &id);
+        let ret: Option<String> = conn.get(key).await?;
+
+        match ret {
+            Some(val) => {
+                let msg: Message = serde_json::from_str(&val)?;
+
+                Ok(Some(msg))
+            }
+            None => Ok(None),
+        }
     }
 
-    async fn run(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn run(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let queue: String = format!("{}.{}", &self.prefix, msg.cmd);
+        let value: String = serde_json::to_string(&msg)?;
+
+        conn.rpush(&queue, &value).await?;
+        conn.ltrim(&queue, 0, self.max_command_messages).await?;
+
+        Ok(())
     }
 
-    async fn forward(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn forward(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let value: String = serde_json::to_string(&msg)?;
+
+        // add to backlog
+        let key: String = format!("{}.backlog.{}", &self.prefix, &msg.uid);
+        conn.set_ex(&key, &value, self.ttl_in_seconds).await?;
+
+        // push to forward for every destination
+        let queue: String = format!("{}.system.forward", &self.prefix);
+        for dst in msg.dst {
+            let value: String = format!("{}.{}", &msg.uid, &dst.to_string());
+            conn.rpush(&queue, &value).await?
+        }
+
+        Ok(())
     }
 
-    async fn reply(&self, _msg: Message) -> anyhow::Result<()> {
-        unimplemented!()
+    async fn reply(&self, msg: Message) -> Result<()> {
+        let mut conn = self.get_connection().await?;
+        let value: String = serde_json::to_string(&msg)?;
```

Comment:
It seems to me that serialization and desrialization from a `String` (or Vec<u8> after you fix the other comments. is  a common operation. Why instead not implement that feature on the message itself.

This can be done in many ways.
- Add custom method `from_bytes` and `to_bytes` method on the Message struct. This is to implement but doesn't look so grate when u going to use it, also doesn't play as nice with other libraries.
- Add trait implementations for standard traits `TryFrom` and `TryInto`. this make it plays more nicely with more things because those traits are std traits hence it can be used in more places.
- Last is to implement the `redis::FromRedisValue` and `redis::ToRedisArgs` this means you can do something like
```rust
conn.rpush(&msg.reg, &msg); // see i am using the message directly
// and
let ret: Option<Message> = conn.get(key).await?; // see i can get the message directly
```

Reply:
I added to_json and from_json (I see they're more specific), will implement FromRedisValue and ToRedisArgs.

Reply:
Added other traits impmenetation (including redis related ones)